File: 백준/Gold/3079. 입국심사/입국심사.cc
#include <iostream>  // stdio.h 와 같은 것
#include <vector>
#include <string>
#include <algorithm>
#include <stack>
#include <queue>
#include <cstring>
#include <cmath>
#include <functional>
#include <map>
#include <unordered_map>
#include <set>
#include <sstream>

// control i
#define MAX 987654321
#define mod1 1000000000
#define mod 1000000007
#define pii pair<int, int>
using ll = long long;
using namespace std;
unsigned long long n,m,k;
int l,r,t;
int h;



vector<ll>v;

int main(){
    
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    
    
    cin >> n >> m;
    // n개의 줄, 사람이 m명
    
    for(int i=0;i<n;i++){
        unsigned long long a;
        cin >> a;
        v.push_back(a);
    }
    
    unsigned long long left = 0;
    unsigned long long right = m * v[0];
    // 심사 기간이 제일 적은 곳에 모든 사람이 검사하는 경우가 최대이다.
    unsigned long long ans = 0;
    
    while(left <= right){
        unsigned long long mid = (left + right) / 2;
        
        unsigned long long num = 0;
        
        for(int i=0;i<n;i++){
            num += ll((mid / v[i]));
        }
        
        if(num >= m){
            if(ans > mid || ans == 0){
                ans = mid;
            }
            right = mid - 1;
        }
        else{
            left = mid + 1;
        }
    }
           
    cout << ans;
    
    
    
    return 0;
    
}

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    
    times.sort()
    
    left = 0
    right = times[len(times) - 1] * n
    # 제일 최악의 경우 모든 손님들이 제일 오래걸리는 사람 한테만 가서 받는 경우 따라서 * n
    
    while left <= right:
        mid = (left + right) // 2
        
        num = 0
        for i in range(len(times)):
            num += (mid // times[i])
            # mid만큼 총 걸릴 경우 times들에 대해서 나눈 값들을 모두 더하면 num이 n이 되어야 한다.
            
        if num < n:
            left = mid + 1
        else:
            if mid <= right:
                answer = mid
            # num이 n이상인데, mid가 아직 right 보다 작을 경우 answer를 넣어준다.
        
            right = mid - 1
    
    return answer